pkg/ingester/errors.go

```diff
@@ -577,3 +595,38 @@ func handlePushErrorWithHTTPGRPC(err error) error {
 	}
 	return err
 }
+
+func handleReadError(err error) error {
+	var (
```

Comment:
Shouldn't we do this?

```suggestion
func handleReadError(err error) error {
	if err == nil {
		return nil
	}
	
	var (
```

I don't see the check in the calls.

Reply:
Wait, I'm wrong because I didn't see that what we call is an ingester function with the same name. Can we just put this code into that method? It's confusing to have a method which sometimes calls a function with same name, IMO.

Reply:
All `handleError`-like functions are placed in `errors.go`, but as the documentation of some of them states, some of them are there only for backwards compatibility and will be removed in 2.12. 
WDYT if we do this kind of improvement once we removed the legacy code?

Reply:
In mimir 2.12 we should get rid of `-ingester.return-only-grpc-errors` too.

Reply:
> All handleError-like functions are placed in errors.go, but as the documentation of some of them states, some of them are there only for backwards compatibility and will be removed in 2.12.

The documentation of this method doesn't state anything. It's just called `handleReadError` and we need to look into it to understand what it's doing. I'd rather call it `mapReadErrorToGRPCStatus`.

OTOH, the caller code is:

```go
func (i *Ingester) handleReadError(err error) error {
	if err == nil {
		return nil
	}

	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
		return err
	}

	if i.cfg.ReturnOnlyGRPCErrors {
		return handleReadError(err)
	}
	return handleReadErrorWithHTTPGRPC(err)
}
```

Which doesn't clarify what is happening (what is _handling_? it should be mapping IMO).

> WDYT if we do this kind of improvement once we removed the legacy code?

I forsee this code staying here for at least between 6 and 9 months, so I would prioritize making it the most expressive and cleaner now.

Reply:
I also noticed that `handleReadError` isn't handling errors, but translating them.

Reply:
I just kept  the same naming convention that was used previously. I will now rename all the methods.

Reply:
I also agree with Oleg that `handleReadError` can be inlined into `Ingester.handleReadError` since it's just cognitive overhead for this tiny amount of code to be in its own function. Another benefit of inlining it would be that the test would be on `ingester.handleReadError` instead, for better coverage.

Reply:
I wouldn't do it right now but only once the HTTP status codes are removed. 
Currently, the goal of `Ingester.HandleReadError()` (renamed in `Ingester.MapReadErrorToErrorWithStatus()`) is to understand whether the `-ingester.return-only-grpc-errors` is true or false, and to call the right function that will do the actual mapping. The corresponding test tests this behavior.

The actual mapping is currently done in `errors.go` (because all the error-related things are in that source), and the tests for both cases of mapping (httpgrpc and grpc) is done there. Once we remove the legacy part, it will make sense to move the mapping in ingester.go. IMO